proof/refine/ARM/Schedule_R.thy

```diff
@@ -3952,6 +3922,302 @@ lemma setNextInterrupt_corres:
+  apply (prop_tac "{((r', s'), r, s). scRefillCount (the (scs_of' s' sc_ptr))
+                                             < scRefillCount (the (scs_of' s sc_ptr))}
+                     = measure (\<lambda>(r, s). scRefillCount (the (scs_of' s sc_ptr)))")
+   apply (clarsimp simp: measure_def inv_image_def split_def)
+  apply (drule sym)
+  apply (erule subst)
+  apply (rule wf_measure)
+  done
+
+lemma mergeRefills_terminates:
+  "sc_at' sc_ptr s' \<Longrightarrow>
+   whileLoop_terminates
+               (\<lambda>_ s'. the (refillHeadOverlapping sc_ptr s'))
+               (\<lambda>_. mergeRefills sc_ptr) r' s'"
+  apply (rule_tac
+           I="\<lambda>_. sc_at' sc_ptr" and
+           R="{((r', s'), (r, s)). scRefillCount (the (scs_of' s' sc_ptr))
+                                    < scRefillCount (the (scs_of' s sc_ptr))}"
+         in whileLoop_terminates_inv)
+    apply simp
+   apply (wpsimp wp: mergeRefills_length_decreasing)
+   apply (fastforce simp: obj_at_simps opt_map_red
+                   dest!: refillHeadOverlapping_implies_count_greater_than_one)
+  apply (rule scRefillCount_wf)
+  done
+
+lemma refillHeadOverlappingLoop_corres:
+  "corres dc (sc_at sc_ptr and is_active_sc2 sc_ptr and pspace_aligned and pspace_distinct)
+             (valid_refills' sc_ptr)
+       (refill_head_overlapping_loop sc_ptr) (refillHeadOverlappingLoop sc_ptr)"
+  unfolding refill_head_overlapping_loop_def refillHeadOverlappingLoop_def runReaderT_def
+  supply refillHeadOverlapping_implies_count_greater_than_one[dest!]
+  apply (rule corres_cross[where Q' = "sc_at' sc_ptr", OF sc_at'_cross_rel], fastforce)
+  apply (rule corres_whileLoop)
+        apply (drule refillHeadOverlapping_corres_eq[where sc_ptr=sc_ptr]; simp add: runReaderT_def)
+       apply simp
+       apply (rule corres_guard_imp
+                      [where P=P and Q=P for P,
+                       where
+                          P'= Q and
+                          Q'="Q and (\<lambda>s. ((\<lambda>sc. 1 < scRefillCount sc) |< scs_of' s) sc_ptr)" for Q])
+         apply (rule corres_cross_add_abs_guard
+                       [where Q="(\<lambda>s. ((\<lambda>sc. 1 < length (sc_refills sc)) |< scs_of2 s) sc_ptr)"])
+          apply clarsimp
+          apply (drule (2) state_relation_sc_relation)
+          apply (clarsimp simp: obj_at_simps is_sc_obj valid_refills'_def sc_relation_def opt_map_red)
+         apply (rule corres_guard_imp)
+           apply (rule mergeRefills_corres)
+          apply simp+
+       apply wpsimp+
+   apply (fastforce simp: obj_at_simps opt_map_red)
+  apply (fastforce intro!: mergeRefills_terminates)
+  done
+
+lemma refillUnblockCheck_corres':
+  "corres dc
+     (sc_at scp and is_active_sc2 scp and pspace_aligned and pspace_distinct
+      and (\<lambda>s. ((\<lambda>sc. 0 < length (sc_refills sc)) |< scs_of2 s) scp))
+     (valid_refills' scp)
+        (refill_unblock_check scp) (refillUnblockCheck scp)"
```

Comment:
We seem to be a bit inconsistent with how we indent the monadic functions in corres statements. My preference would be
```suggestion
  "corres dc
     (sc_at scp and is_active_sc2 scp and pspace_aligned and pspace_distinct
      and (\<lambda>s. ((\<lambda>sc. 0 < length (sc_refills sc)) |< scs_of2 s) scp))
     (valid_refills' scp)
     (refill_unblock_check scp) (refillUnblockCheck scp)"
```

Reply:
I seem to like having the functions not aligned with the preconditions, but no strong preference.